### PR TITLE
[CLOUD] Upload images to CSP

### PR DESCRIPTION
**Pull request by cfconrad**
Upload images to CSP either azure or amazon ec2.
The image, which is given by PUBLIC_CLOUD_IMAGE_LOCATION, gets
downloaded first and then uploaded to CSP.

Extract common functions in a helper classes. This
can be later extended, e.g. calling IPA tool.
Long time goal, is to replace the cli calls by native perl api
calls like paws and azure-sdk-perl.

The upload_image test module exit successful, if a image or
resource group (azure) with the same name already exists.
Once the image is uploaded or the image exists on CSP the
variable PUBLIC_CLOUD_IMAGE_ID is set with the proper value.

Add file ec2utils.conf, it is used by ec2uploadimg.

- Verification run:
   azure: http://cfconrad-vm.qa.suse.de/tests/697
   aws ec2: http://cfconrad-vm.qa.suse.de/tests/696 (image exists)
   aws ec2: http://cfconrad-vm.qa.suse.de/tests/699

### CODE REVIEW

File: tests/publiccloud/upload_image.pm [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    else {
+        die('Unknown PUBLIC_CLOUD_PROVIDER given');
+    }
+}
+
+sub run {
+    my ($self) = @_;
+    select_virtio_console();
+
+    prepare_os;
```
Comment (asmorodskyi) [inline tests/publiccloud/upload_image.pm]: `prepare_os();`

File: lib/publiccloud/ec2.pm [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+          . "--machine 'x86_64' "
+          . "-n '" . $self->prefix . '-' . $img_name . "' "
+          . (($img_name =~ /hvm/i) ? "--virt-type hvm --sriov-support " : "--virt-type para ")
+          . "--verbose "
+          . "--regions '" . $self->region . "' "
+          . "--ssh-key-pair '" . $self->ssh_key . "' "
+          . "--private-key-file " . $self->ssh_key_file . " "
+          . "-d 'OpenQA tests' "
+          . "'$file'",
+        timeout => 60 * 60
```
Comment (asmorodskyi) [inline lib/publiccloud/ec2.pm]: I vote for defining parent class parameter `timeout` and give value in each children

File: lib/publiccloud/provider.pm [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+=head1 METHODS
+ 
+=head2 init
+ 
+Needs provider specific credentials, e.g. key_id, key_secret, region.
+
+=cut
+sub init {
+    die('init() isn\'t implemented');
```
Comment (asmorodskyi) [inline lib/publiccloud/provider.pm]: would say something like "Can not init base class"

File: lib/publiccloud/provider.pm [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+
+=head2 find_img
+
+Retrieves the image-id by given image C<name>.
+
+=cut
+sub find_img {
+    die('find_image() isn\'t implemented');
```
Comment (asmorodskyi) [inline lib/publiccloud/provider.pm]: I am not sure here , but isn't this code will be never ever executed ? I mean you already dying in int or as usual there always a way to do it if you really want ?

File: tests/publiccloud/upload_image.pm [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        return;
+    }
+
+    assert_script_run("wget $img_url -O $img_name", timeout => 60 * 10);
+
+    my $img_id = $provider->upload_img($img_name);
+
+    set_var('PUBLIC_CLOUD_IMAGE_ID', $img_id);
+
+    $provider->cleanup();
```
Comment (asmorodskyi) [inline tests/publiccloud/upload_image.pm]: the fact that function not implemented for ec2 will not fail the code when executed for ec2 ?

File: lib/publiccloud/ec2.pm (hunk 2 above)
Comment (cfconrad) [reply lib/publiccloud/ec2.pm]: I'm not sure. The timeout depends on the provider and how images gets uploaded. I do also not see to define a generic timeout, when doing so, then it should be a specific upload_image_timeout.
What I see here is, give the timeout as parameter to the upload image function and if not given the default from the specific implementation.

File: lib/publiccloud/provider.pm (hunk 3 above)
Comment (cfconrad) [reply lib/publiccloud/provider.pm]: Yes, a more meaningful message makes more sense, thx.

File: lib/publiccloud/provider.pm (hunk 4 above)
Comment (cfconrad) [reply lib/publiccloud/provider.pm]: The intention was, if someone use a implementation (e.g. google cloud), which doesn't provide this method, will be informed. It's like a abstract method.
I could just return undef, so a caller would never get a image. But then I do not have a difference between not implemented and image not found.
I would vote for that not implemented approach.

File: tests/publiccloud/upload_image.pm (hunk 5 above)
Comment (cfconrad) [reply tests/publiccloud/upload_image.pm]: The method is not implement for azure, but then the method of the base class is called which is just a empty one.
So if the derived class doesn't need to cleanup something, just nothing happen.

File: lib/publiccloud/ec2.pm (hunk 2 above)
Comment (asmorodskyi) [reply lib/publiccloud/ec2.pm]: you misunderstood my idea, but let's skip it for now

**Comment by asmorodskyi** [comment]
really awesome PR !

**Comment by cfconrad** [comment]
thx :-)

**Pull request merged**